cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,1164 @@
+
+    thrust::for_each(handle.get_thrust_policy(),
+                     thrust::make_counting_iterator<edge_t>(0),
+                     thrust::make_counting_iterator<edge_t>(intersection_indices.size()),
+                     extract_p_r<vertex_t, edge_t, decltype(edges)>{
+                     num_edges,
+                     raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+                     raft::device_span<vertex_t const>(intersection_indices.data(), intersection_indices.size()),
+                     raft::device_span<vertex_t>(num_triangles.data(), num_triangles.size()),
+                     edges});
+    
+    thrust::for_each(handle.get_thrust_policy(),
+                     thrust::make_counting_iterator<edge_t>(0),
+                     thrust::make_counting_iterator<edge_t>(intersection_indices.size()),
+                     extract_q_r<vertex_t, edge_t, decltype(edges)>{
+                     num_edges,
+                     raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+                     raft::device_span<vertex_t const>(intersection_indices.data(), intersection_indices.size()),
+                     raft::device_span<vertex_t>(num_triangles.data(), num_triangles.size()),
+                     edges});
+
+    auto edges_to_num_triangles = thrust::make_zip_iterator(
+      edges, num_triangles.begin());
+
+    // 'invalid_edge_first' marks the beginning of the edges to be removed
+    auto invalid_edge_last =
+      thrust::stable_partition(handle.get_thrust_policy(),
+                               edges_to_num_triangles,
+                               edges_to_num_triangles + num_triangles.size(),
+                               [k] __device__(auto e) {
+                                 auto num_triangles = thrust::get<1>(e);
+                                 return num_triangles < k - 2;
+                               });
+    
+    size_t num_invalid_edges{0};
+    num_invalid_edges = static_cast<size_t>(
+      thrust::distance(edges_to_num_triangles, invalid_edge_last));
+
+
+    // copy invalid edges
+    auto invalid_edges_buffer = allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(
+      num_invalid_edges, handle.get_stream());
+    
+    thrust::copy(handle.get_thrust_policy(),
+                 edges,
+                 edges + num_invalid_edges,
+                 get_dataframe_buffer_begin(invalid_edges_buffer));
+    
+    // sort back the edges as those are needed later when running a binary tree
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        edges,
+                        edges + num_edges,
+                        num_triangles.begin());
+
+
+    // Unroll and remove/mask edges as long as there are still edges part
+    // of the K-Truss.
+    auto num_valid_edges = num_edges - num_invalid_edges;
+    while ((num_valid_edges != 0) && (num_invalid_edges !=0)) {
```

Comment:
Where is case 1?

Reply:
And we should assume that intersection_offsets & intersection indices won't be available for the entire edge list at this point.

intersection_indices.size() can far exceed the number of edges. So, if we need to process edges in chunks to compute edge triangle counts. Once we update edge triangle count to consider this, we should assume intersection_offsets & intersection_indices won't be available at this moment.

The first thing we need to do here is to re-compute intersection_offsets & intersection_indices but only for the edges to be deleted.

Then, using this information, we can perform case 1: unroll (p, q).

Reply:
Case 1 is there. Also I will apply nbr intersection in chunks once I confirm the implementation is correct as it is easier to debug